File: server/validation.py
"""
Validation module for StudyBuddy backend.

This module provides centralized validation logic for session data,
time validation, and file uploads.
"""

from datetime import datetime, timezone
from typing import Dict, Any, Optional


class ValidationError(Exception):
    """Custom exception for validation errors.
    
    Attributes:
        field: The field name that failed validation
        message: Human-readable error message
    """
    
    def __init__(self, field: str, message: str):
        self.field = field
        self.message = message
        super().__init__(f"{field}: {message}")


def validate_session_time(time_str: str) -> bool:
    """Validate that session time is in the future.
    
    Args:
        time_str: ISO format datetime string (e.g., "2024-01-15T19:00:00Z")
        
    Returns:
        True if validation passes
        
    Raises:
        ValidationError: If time is not in the future
    """
    try:
        # Parse ISO format time string, handle both 'Z' and '+00:00' UTC indicators
        session_time = datetime.fromisoformat(time_str.replace('Z', '+00:00'))
        now = datetime.now(timezone.utc)
        
        if session_time <= now:
            raise ValidationError('time', 'Session time must be in the future')
        
        return True
    except ValueError as e:
        raise ValidationError('time', f'Invalid time format: {str(e)}')


def validate_session_data(data: Dict[str, Any]) -> bool:
    """Validate all required session fields.
    
    Args:
        data: Dictionary containing session data
        
    Returns:
        True if all validations pass
        
    Raises:
        ValidationError: If any required field is missing or invalid
    """
    required_fields = ['course', 'subject', 'building', 'time', 'duration']
    
    # Check for required fields
    for field in required_fields:
        if not data.get(field):
            raise ValidationError(field, f'{field} is required')
    
    # Validate session time is in the future
    validate_session_time(data['time'])
    
    return True


def validate_image_file(file_data: Any, max_size_mb: int = 5) -> bool:
    """Validate profile picture file.
    
    Args:
        file_data: File object or file-like object with read() and content_type
        max_size_mb: Maximum file size in megabytes (default: 5MB)
        
    Returns:
        True if validation passes
        
    Raises:
        ValidationError: If file is invalid
    """
    # Check if file exists
    if not file_data:
        raise ValidationError('file', 'No file provided')
    
    # Check file type
    allowed_types = ['image/jpeg', 'image/jpg', 'image/png', 'image/gif', 'image/webp']
    content_type = getattr(file_data, 'content_type', None)
    
    if content_type and content_type not in allowed_types:
        raise ValidationError('file', 'Please select a valid image file (JPEG, PNG, GIF, or WebP)')
    
    # Check file size
    if hasattr(file_data, 'read'):
        # Read file to check size
        file_data.seek(0)
        file_content = file_data.read()
        file_size = len(file_content)
        file_data.seek(0)  # Reset file pointer
        
        max_size_bytes = max_size_mb * 1024 * 1024
        if file_size > max_size_bytes:
            raise ValidationError('file', f'Image must be less than {max_size_mb}MB')
    
    return True

File: client/src/utils/validation.js
/**
 * Centralized validation utilities for StudyBuddy frontend
 * Provides client-side validation for user input
 * Note: Backend validation is authoritative - these are for UX only
 */

/**
 * Validate that a session time is in the future
 * @param {string|Date} timeValue - The time to validate (ISO string or Date object)
 * @returns {Object} { valid: boolean, error: string|null }
 */
export function validateSessionTime(timeValue) {
  if (!timeValue) {
    return { valid: false, error: 'Session time is required' }
  }

  const sessionTime = new Date(timeValue)
  const now = new Date()

  if (isNaN(sessionTime.getTime())) {
    return { valid: false, error: 'Invalid date format' }
  }

  if (sessionTime <= now) {
    return { valid: false, error: 'Please choose a future date and time' }
  }

  return { valid: true, error: null }
}

/**
 * Validate all required session data fields
 * @param {Object} data - Session data object
 * @param {string} data.course - Course/topic name
 * @param {string} data.subject - Subject area
 * @param {string} data.building - Building location
 * @param {string} data.time - Session time (ISO string)
 * @param {string} data.duration - Session duration
 * @returns {Object} { valid: boolean, errors: Object }
 */
export function validateSessionData(data) {
  const errors = {}

  // Required fields validation
  if (!data.course || !data.course.trim()) {
    errors.course = 'Course/topic is required'
  }

  if (!data.subject || !data.subject.trim()) {
    errors.subject = 'Subject area is required'
  }

  if (!data.building || !data.building.trim()) {
    errors.building = 'Building is required'
  }

  if (!data.time) {
    errors.time = 'Session time is required'
  } else {
    // Validate time is in future
    const timeValidation = validateSessionTime(data.time)
    if (!timeValidation.valid) {
      errors.time = timeValidation.error
    }
  }

  if (!data.duration || !data.duration.trim()) {
    errors.duration = 'Duration is required'
  }

  return {
    valid: Object.keys(errors).length === 0,
    errors
  }
}

/**
 * Validate image file for profile picture upload
 * @param {File} file - The file to validate
 * @returns {Object} { valid: boolean, error: string|null }
 */
export function validateImageFile(file) {
  if (!file) {
    return { valid: false, error: 'Please select a file' }
  }

  // Check file type
  if (!file.type.startsWith('image/')) {
    return { valid: false, error: 'Please select a valid image file' }
  }

  // Check file size (5MB limit)
  const maxSize = 5 * 1024 * 1024 // 5MB in bytes
  if (file.size > maxSize) {
    return { valid: false, error: 'Image must be less than 5MB' }
  }

  return { valid: true, error: null }
}

/**
 * Validate password for password reset
 * @param {string} password - The password to validate
 * @returns {Object} { valid: boolean, error: string|null }
 */
export function validatePassword(password) {
  if (!password) {
    return { valid: false, error: 'Password is required' }
  }

  if (password.length < 8) {
    return { valid: false, error: 'Password must be at least 8 characters' }
  }

  return { valid: true, error: null }
}

/**
 * Validate email format
 * @param {string} email - The email to validate
 * @returns {Object} { valid: boolean, error: string|null }
 */
export function validateEmail(email) {
  if (!email) {
    return { valid: false, error: 'Email is required' }
  }

  const emailRegex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/
  if (!emailRegex.test(email)) {
    return { valid: false, error: 'Please enter a valid email address' }
  }

  return { valid: true, error: null }
}

/**
 * Validate required field
 * @param {any} value - The value to validate
 * @param {string} fieldName - Name of the field for error message
 * @returns {Object} { valid: boolean, error: string|null }
 */
export function validateRequired(value, fieldName = 'This field') {
  if (!value || (typeof value === 'string' && !value.trim())) {
    return { valid: false, error: `${fieldName} is required` }
  }

  return { valid: true, error: null }
}
